keps/sig-auth/2784-csr-duration/README.md

```diff
@@ -0,0 +1,473 @@
+2. Not upgraded (or partially upgraded) controller manager
+3. Upgraded client
+
+In this scenario, the requested `spec.expirationSeconds` may be ignored because
+the controller manger will not understand this field.  This is harmless and
+represents the status quo.
+
+#### Scenario 2
+
+1. Partially upgraded API server
+2. Upgraded controller manager
+3. Upgraded client
+
+In this scenario, the requested `spec.expirationSeconds` may be ignored because
+old API servers will silently drop the field on update.  This is harmless
+represents and the status quo.
+
+The CSR API is resilient to split brain scenarios as unknown fields are silently
+dropped and the `spec` fields are immutable after creation [1] [2] [3].
+
+[1]: https://github.com/kubernetes/kubernetes/blob/24b716673caae31f070b06a337bc12c97ff1d4cb/pkg/registry/certificates/certificates/strategy.go#L104-L112
+[2]: https://github.com/kubernetes/kubernetes/blob/24b716673caae31f070b06a337bc12c97ff1d4cb/pkg/registry/certificates/certificates/strategy.go#L175-L176
+[3]: https://github.com/kubernetes/kubernetes/blob/24b716673caae31f070b06a337bc12c97ff1d4cb/pkg/registry/certificates/certificates/strategy.go#L297-L298
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- Other
+  - Describe the mechanism:
+      As written above, this functionality will be enabled by default with no configuration options.
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+      N/A
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+      N/A
+
+###### Does enabling the feature change any default behavior?
+
+Existing clients would continue to leave `spec.expirationSeconds` unset and thus
+would observe no difference in behavior.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+N/A
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+N/A
+
+###### Are there any tests for feature enablement/disablement?
+
+Unit and integration tests will cover cases where `spec.expirationSeconds` is
+specified and cases where it is left unspecified.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Since it is optional for signers to honor `spec.expirationSeconds`, this design
+is fully tolerant of API server and controller manager rollouts/rollbacks that
+fail or get wedged in a partial state.  The worse case scenario is that the
+`spec.expirationSeconds` field is ignored, which mimics the status quo.  Clients
+must always check the duration of the issued certificate to determine if the
+requested `spec.expirationSeconds` was honored.
+
+###### What specific metrics should inform a rollback?
+
+N/A
```

Comment:
do we want visibility to clients requesting expirations that get truncated?

Reply:
Not sure if such a metric would be useful because there is no way for a client to know if the requested expiration is aligned with the certificate controller config.  Since we do not have metrics related to the CSR controllers today so I left this out.

Reply:
Without a metric, how will a cluster-admin know that this feature is functioning?  How will he know that he has approvers that are not honoring the value in the long term?

It seems like a thing worth tracking.

Reply:
Added explicit metrics around this.